### PR TITLE
rego: make wasmtime-go dependency "more optional"

### LINKED ISSUE #3545: wasmtime dependency includes large binary file
Sometime between v0.24.0 and v0.29.4, we added a dependency on `github.com/bytecodealliance/wasmtime-go`. There's some history around why we used this particular version, but it boils down to that it has an active community and has everything we need.

The downside is that the repository includes a large binary file which `go mod vendor` adds to `vendor/` (See: vendor/github.com/bytecodealliance/wasmtime-go/build/linux-x86_64/libwasmtime.a). GitHub complains about this, but it doesn't appear to have any other negative effects.

We'd rather not require all repositories which vendor open-policy-agent/opa to download this (possibly unnecessary for us?) file, but there doesn't seem to be a clean solution on our end. The fix will probably either come down to one of:

- convincing bytecodealliance/wasmtime to not include the binary (or figuring out how to work with go modules to only download it if we actually depend on it)
- finding a way to get `go mod vendor` to ignore the binary on our end
- using a different implementation

Again, there isn't a pressing need and nothing is blocked or broken because of this.

### PR DESCRIPTION
**Pull request by srenatus**
Users of OPA as a library are concerned about big binary blobs in their vendor directories. Even more so if they don't use them. This is the case for anyone using OPA as library, but not using the wasm-backed evaluation feature.

Fixes #3545.

--------

With this change, importers of any packages **other than `server` and `cmd`** will have to explicitly opt-in to using wasm evaluation features by having an underscore import somewhere,
```go
import _ "github.com/open-policy-agent/opa/features/wasm"
```
~besides passing the `opa_wasm` build tag to your `go build` call.~

------

Things up for debate, or further todos:

1. This is a breaking change. ⚡ Before, enabling wasm-support only required passing the build tag; now, it requires another source change.

   Do we want to go that route, do the benefits outweigh that?
2. Can we drop the build tag now?

    Right now, it's mainly used to build a static linux binary without wasm enabled from the same source checkout. I think we could drop the tag from all but `cmd/features.go` and `server/features.go` and achieve the same result. That way, we would remove the second requirement above -- anyone underscore-importing the `features/wasm` package would opt into using wasm, but that one change alone.

### CODE REVIEW

**Comment by maxsmythe** [comment]
Do we know the scope of impact for the breaking change? Are there any users of the OPA library that compile to WASM other than the OPA CLI? Any way to figure this out via github search?

It does sound like the build tag is redundant if the consumer already has control over whether the WASM binaries are included.

**Comment by srenatus** [comment]
> Do we know the scope of impact for the breaking change? Are there any users of the OPA library that compile to WASM other than the OPA CLI? Any way to figure this out via github search?

To be clear, this is about _evaluating Wasm_ -- compiling a policy to wasm doesn't depend on any of wasmtime-go.

It would affect those users using the rego package with the `rego.Target("wasm")` option in `rego.New()` and I believe nobody else: the code backing the rego packages is all internal, `github.com/open-policy-agent/opa/internal/rego/opa` ➡️ `github.com/open-policy-agent/opa/internal/wasm/sdk/opa`. I believe that those aren't too many. Also, with clear instructions, it might be forgivable.

**Comment by tsandall** [comment]
> This is a breaking change. 

In this case, since the # of users currently relying on the internal wasm runtime (not the compiler! just the runtime...) is relatively low (compared to broader library usage), I'm fine w/ the breaking change. Let's just make sure to call it out _very_ clearly in the CHANGELOG.

> Can we drop the build tag now?

Just so that I understand... you're suggesting that we remove the `+build opa_wasm` constraint from everywhere except `cmd/features.go` and `server/features.go`. Library users that want the wasm runtime available can include the feature import. Builds that need to produce a wasm-enabled OPA or a wasm-disabled OPA do not have to change (i.e., behaviour of make targets won't change and users running `go build` themselves won't be affected.)

@srenatus I'm curious why we need `servers/features.go`.

**Comment by srenatus** [comment]
> > Can we drop the build tag now?
> 
> Just so that I understand... you're suggesting that we remove the +build opa_wasm constraint from everywhere except cmd/features.go and server/features.go. Library users that want the wasm runtime available can include the feature import. Builds that need to produce a wasm-enabled OPA or a wasm-disabled OPA do not have to change (i.e., behaviour of make targets won't change and users running go build themselves won't be affected.)

Exactly. The build tag's name is a bit unfortunate then, because it's not really about "opa_wasm" -- that's controlled differently. It's more like "static" where enabling "static build" means that some features won't be supported. I think we could even put the `+build !static` into `features/wasm.go` then.

_But_ that'll be a potentially breaking change for manual build commands. ⚡ I guess we could just keep it as-is -- a little of the mark, but for a good reason.

> @srenatus I'm curious why we need servers/features.go.

The rationale here was that running wasm bundles would have worked before if you'd use the server package, and would then potentially break. That bit was done in haste, however, I think we could also accept the breaking change _here_, and add the feature import to the test files exercising that behaviour.

**Review by tsandall — APPROVED** [review]

File: rego/rego.go [hunk 1]
```diff
@@ -1501,7 +1502,21 @@ func (r *Rego) PrepareForEval(ctx context.Context, opts ...PrepareOption) (Prepa
...
+				fmt.Fprintf(os.Stderr, `WebAssembly runtime not supported in this build.
+----------------------------------------------------------------------------------
+Please download an OPA binary with Wasm enabled from
+  https://www.openpolicyagent.org/docs/latest/#running-opa
+or build it yourself (with Wasm enabled).
+----------------------------------------------------------------------------------
+`)
+				os.Exit(1)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (srenatus) [reply rego/rego.go:1517]: We should not. I'll fix that. I wasn't really happy with the current behaviour either 😄 (i.e. the os.Exit in the no-op implementation)

**Comment by srenatus** [comment]
I'm inclined to get this merged. We can, if need be, still do adjustments.

Since our last deeper wasmtime-build changes, we've gotten alpine-image based tests for the static binaries into CI, and run the binaries on all platforms... we should be safe here.

I've manually verified that conftest would -- if they _did_ vendor (they don't) -- no longer carry wasmtime blobs, and neither would gatekeeper.

**Pull request merged**